examples/cosmos/ics20/ics20.qnt

```diff
@@ -112,26 +128,83 @@ module ics20 {
 
     pure val newChainState = chainState.with("bank", bankResult.accounts)
     pure val ack = if (bankResult.success) {
-      { success: true, errorMessage: "" }
+      { success: true, errorMessage: "", packet: packet }
     } else {
-      { success: false, errorMessage: errorMessage }
+      { success: false, errorMessage: errorMessage, packet: packet }
     }
 
     (newChainState, ack)
   }
+
+  /// Refund tokens from a token transfer initiated by `packet`.
+  ///
+  /// Should be called if the token transfer fails on the receiver chain
+  /// (due to a packet acknowledgement indicating failure, or due to a timeout).
+  ///
+  /// In particular, `refundTokens` is called by both `onAcknowledgePacket` (on
+  /// failure) and by `onTimeoutPacket`, to refund burnt or escrowed tokens to
+  /// the original sender.
+  ///
+  /// `packet`: The packet (originally sent from the local chain) that initiated the token transfer.
+  ///           Packet `packet` should have resulted in a failed acknowledgement or timeout.
+  pure def refundTokens(chainState: ChainState, packet: Packet): ChainState = {
+    pure val data: FungibleTokenData = packet.data
+    pure val source: HopElement = { port: packet.sourcePort, channel: packet.sourceChannel }
+    pure val bankResult =
+      if (movingBackAlongTrace(data.denom, source)) {
+        // The (failed) transfer attempt was moving the token back along its denom trace.
+        // For this, we originally burnt vouchers in `sendFungibleTokens`, so mint them back.
+        chainState.bank.MintCoins(data.sender, data.denom, data.amount)
+      } else {
+        // The (failed) transfer attempt was moving the token further along its denom trace.
+        // For this, we originally escrowed tokens in `sendFungibleTokens`, so unescrow them back.
+        pure val escrowAccount = chainState.channelEscrowAddresses.get(packet.srcChannel)
+        chainState.bank.TransferCoins(escrowAccount, data.sender, data.denom, data.amount)
+      }
+    if (bankResult.success) {
+      chainState.with("bank", bankResult.accounts)
+    } else {
+      // TODO: ICS 20 does not specify what happens if the bank modules return an error.
```

Comment:
We shouldn't forget about that. Shall we open an issue on the IBC spec repo?

Reply:
Will do! I've filed https://github.com/informalsystems/apalache-team-services/issues/14 to follow up on.